hedging/src/okex_transfers/mod.rs

```diff
@@ -0,0 +1,206 @@
+use rust_decimal::Decimal;
+use sqlx::{Executor, PgPool};
+use uuid::Uuid;
+
+use okex_client::{ClientTransferId, TransferState};
+use shared::pubsub::CorrelationId;
+
+use crate::error::HedgingError;
+
+pub struct ReservationSharedData {
+    pub correlation_id: CorrelationId,
+    pub action_type: String,
+    pub action_unit: String,
+    pub target_usd_exposure: Decimal,
+    pub current_usd_exposure: Decimal,
+    pub trading_btc_used_balance: Decimal,
+    pub trading_btc_total_balance: Decimal,
+    pub current_usd_btc_price: Decimal,
+    pub funding_btc_total_balance: Decimal,
+}
+
+pub struct Reservation<'a> {
+    pub action_size: Option<Decimal>,
+    pub transfer_type: String,
+    pub fee: Decimal,
+    pub transfer_from: String,
+    pub transfer_to: String,
+    pub shared: &'a ReservationSharedData,
+}
+
+#[derive(Clone)]
+pub struct OkexTransfers {
+    pool: PgPool,
+}
+
+impl OkexTransfers {
+    pub async fn new(pool: PgPool) -> Result<Self, HedgingError> {
+        Ok(Self { pool })
+    }
+
+    pub async fn reserve_transfer_slot<'a>(
+        &self,
+        reservation: Reservation<'a>,
+    ) -> Result<Option<ClientTransferId>, HedgingError> {
+        let mut tx = self.pool.begin().await?;
+        tx.execute("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+            .await?;
+        let res = sqlx::query!(
+            r#"SELECT client_transfer_id FROM okex_transfers WHERE state = 'pending' AND lost = false AND transfer_type = $1"#,
+            reservation.transfer_type,
+        )
+        .fetch_all(&mut tx)
+        .await?;
+
+        if !res.is_empty() {
+            return Ok(None);
+        }
+        let id = ClientTransferId::new();
+        sqlx::query!(
+            r#"INSERT INTO okex_transfers (
+                client_transfer_id, 
+                correlation_id, 
+                action, 
+                transfer_type, 
+                currency,
+                amount,
+                fee,
+                transfer_from,
+                transfer_to,
+                target_usd_exposure,
+                current_usd_exposure,
+                trading_btc_used_balance,
+                trading_btc_total_balance,
+                current_usd_btc_price,
+                funding_btc_total_balance,
+                state
+            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)"#,
+            String::from(id.clone()),
+            Uuid::from(reservation.shared.correlation_id),
+            reservation.shared.action_type,
+            reservation.transfer_type,
```

Comment:
I'm not sure we should have 2 slots for the 2 different transfer types. Just having 1 slot will simplify the logic.